File: cpp/src/player.cpp
/******************************************************************************
* AUTHOR:      Alexander Casal
* FILE:        player.cpp
* DESCRIPTION: In this case, Player is the Client. This is where we need the
*              ability to select different strategies at runtime. The Player
*              equips a weapon. This weapon can be exchanged for another at
*              any time. When a weapon is equipped, the player can attack
*              enemies with it.
*****************************************************************************/
#include "player.h"
#include <iostream>

 /**
 * Constructor
 *
 * At first, Player starts without any weapon equipped.
 */
Player::Player()
{
	weapon = nullptr;
}

/**
* changeWeapon
*
* Provide the ability to change strategies (Weapon) at runtime.
*/
void Player::changeWeapon(std::unique_ptr<Weapon> weapon)
{
	this->weapon = std::move(weapon);
}

/**
* attack
*
* Allow the Player to attack using the assigned strategy (Weapon).
*/
void Player::attack()
{
	if (weapon != nullptr)
	{
		weapon->attack();
	}
	else
	{
		std::cout << "You don't have a weapon equipped!\n";
	}
}

File: java/Player.java
package strategygame;

/**
 * The Player equips a weapon. This weapon can be exchanged for another
 * at any time. When a weapon is equipped, the player can attack enemies
 * with it.
 * @author Alexander
 */
public class Player {
    
    private Weapon weapon;
    
    /**
     * Constructor
     * 
     * Start off without any weapon equipped
     */
    public Player() {
        
        weapon = null;
    }
    
    /**
     * changeWeapon
     * 
     * Changes the weapon being used by the player
     * @param weapon The new weapon
     */
    public void changeWeapon(Weapon weapon) {
        
        this.weapon = weapon;
    }
    
    /**
     * attack
     * 
     * Allow the player to attack using his/her currently equipped weapon
     */
    public void attack() {
        
        if (weapon != null) {
            
            weapon.attack();
        } else {
            
            System.out.println("You don't have a weapon equipped!");
        }
    }
}
